core/src/main/java/io/grpc/Metadata.java

```diff
@@ -222,8 +283,17 @@ public boolean containsKey(Key<?> key) {
    *
    * @return unmodifiable Set of keys
    */
+  @SuppressWarnings("deprecation") // The String ctor is deprecated, but fast.
   public Set<String> keys() {
-    return Collections.unmodifiableSet(store.keySet());
+    if (size == 0) {
+      return Collections.emptySet();
+    }
+    Set<String> ks = new HashSet<String>(size);
```

Comment:
Maybe, but not sure if we want to commit to it.  The removeAll function is a place where I wish we would not have added it, since /we/ don't even use.  

Someone (I can't remember who) asked for this function but didn't give a lot of requirements.  If people care we can always add another method in later.